### PR TITLE
Thumbnail scrolling

### PR DESCRIPTION
**Pull request by will-moore**
See https://trello.com/c/JLpfTbmW/4-improve-thumbnails-loading-display

Instead of clicking up/down to load more thumbnails, the thumbnail panel is created at the full size (with only the visible thumbnails loaded) and more thumbnails are loaded when we scroll.
There are lots of different cases to handle and a lot has changed so everything needs testing:

 - Expected behaviour: any thumbnails that appear in the panel should be loaded (no empty grey squares).
 - While actually scrolling we don't load thumbnails, since that would load the whole dataset if you scroll fast from one end to the other. Instead, wait until scrolling slows / stops, then load thumbs for visible thumbnails.
 - You can see when loading is happening in the devtools Network tab.
 - In all scenarios below, check thumbs on load, then scroll about, then refresh thumbs (when scrollbar is in various locations).
 - Check that IE works.

**Scenarios to test:**

 - Single image in 1 or more Datasets, double-clicked in webclient. url like ```/webclient/img_detail/4413/?dataset=1301```
        - Check that thumbs loaded from *current* parent Dataset (not a different Dataset that the image is in)
        - Try double-clicking images in webclient near the start of the Dataset and at the end. Thumb panel in iviewer should scroll to the show the selected image.
        - Try with a small Dataset (< 5 images) and as large as possible.
        - Remove the ```?datasets=ID``` from the URL and reload. If image is in single Dataset, thumbs should be loaded. If more than 1 Dataset, thumbnail panel hidden.
 - Multiple selected images > Open with iviewer. Url like ```/iviewer/?images=4402,4403,4404,4405 ```
 - Orphaned image, double-clicked in webclient. e.g. ```webclient/img_detail/57012/``` should hide thumbnail panel.
 - Well - openwith > iviewer.  Url ```/iviewer/?well=148``` First image should be selected.
 - Datasets - openwith > iviewer. ```/iviewer/?dataset=10480``` First image selected.

### CODE REVIEW

**Comment by mtbc** [comment]
I suspect that dropping IE11 (or not) should be a clear, conscious decision given that we often choose to support old stuff.

**Comment by jburel** [comment]
Basic testing:
* Open dataset with < 10 Images by double-clicking on an image. The correct thumbnails were loaded
* Open a larger dataset (~80 images) visible thumbnails loaded. Scroll down to the bottom: more visible thumbnails are then loaded. Moving backward forward to load the rest of the thumbnails

**Comment by jburel** [comment]
``
- Remove the ?datasets=ID from the URL and reload. If image is in single Dataset, thumbs should be loaded. If more than 1 Dataset, thumbnail panel hidden.
- Datasets - openwith > iviewer. /iviewer/?dataset=10480 First image selected.
- Well - openwith > iviewer. Url /iviewer/?well=148 First image should be selected.
- Orphaned image, double-clicked in webclient. e.g. webclient/img_detail/57012/ should hide thumbnail panel.
This works fine
``
Multiple selected images > Open with iviewer. Url like /iviewer/?images=4402,4403,4404,4405
``
If I select all the images in the dataset All the thumbnails are loaded
but If I open the dataset or double click on a thumbnail to open the viewer. Only few thumbnails are loaded.

**Comment by jburel** [comment]
I don't think we can drop IE

**Comment by jburel** [comment]
Safari:
* Clicking on well with only one field does not hide the thumbnails panel. It is empty
pr-207_safari

**Comment by jburel** [comment]
Noticeable loading difference between FF and chrome 
In FF I have time to see "Not Found"
even for small dataset

**Comment by jburel** [comment]
FF 
Same problem as the one described in https://github.com/ome/omero-iviewer/pull/207#issuecomment-421030884

This time
it does not work for well with multiple fields.
pr-2017-ff

**Comment by mtbc** [comment]
I too saw that "Not Found": https://openmicroscopy.slack.com/archives/C0K5EED3R/p1535619171000100

**Comment by jburel** [comment]
kymograph is correctly displayed cc @pwalczysko 
pr-207-kymo

**Comment by jburel** [comment]
While testing 5.4.8 
Looking at data belonging to another user when that user has not viewed the image
leads to "Not found" in the left-hand panel and empty bird eye view
we should handle such case a bit better

**Comment by will-moore** [comment]
@jburel Re: "If I select all the images in the dataset All the thumbnails are loaded
but If I open the dataset or double click on a thumbnail to open the viewer. Only few thumbnails are loaded."
As long as all the thumbnails that are *visible* are loaded then I think this is OK?
I assume that the number of images listed with ?image=1,2,3 is not going to be larger than a safe number of thumbnails to load (e.g. 200, or even more).

**Comment by jburel** [comment]
All the thumbnails were loaded even the non visible when I do a selection of images
I probably select ~100

**Comment by will-moore** [comment]
Also noticed that e.g. http://web-dev-merge.openmicroscopy.org/iviewer/?well=339085 (user-1) doesn't work (no thumbnails or image loaded).

**Comment by will-moore** [comment]
Re: https://github.com/ome/omero-iviewer/pull/207#issuecomment-421319766 The 'Not Found' in this case is the same as previously (not related to this PR). To fix that we could use /webclient/render_thumbnail/ID instead of /webgateway/render_thumbnail. Then we'd get the 'No Image'  icon same as webclient?

**Comment by will-moore** [comment]
Fixed the placeholder "Not Found" while images loading in Firefox.
Also fixed bugs with SPW images in a Well (```/webclient/img_detail/ID/```) and opening Well directly (```/iviewer/?well=ID```). In both cases, the thumbnails within the Well should be shown in the thumbnail_slider, even if it's only 1 field (same behaviour as before).

File: src/app/thumbnail-slider.js [hunk 1]
```diff
@@ -700,6 +785,15 @@ export default class ThumbnailSlider extends EventSubscriber {
             }, 2000);
     }
 
+    setThumbnailsCount(count) {
+        console.log('setThumbnailsCount:', count, 'was:', this.thumbnails.length);
+        if (typeof count !== 'number') {
+            count = 0;
+        }
+        this.thumbnails.length = count;
+        this.thumbnails.fill({'version': 0, 'title': 'unloaded'});
```
Comment (mtbc) [inline src/app/thumbnail-slider.js]: Given https://github.com/ome/omero-iviewer/pull/207#issuecomment-421028262 is this okay?

**Comment by jburel** [comment]
* Wells with 1 or multiple fields are now correctly displayed in FF,  Vivaldi and Safari
* I also select a large number of images, loading occurred when scrolling down the list

No cover yet: IE testing

File: src/app/thumbnail-slider.js [hunk 2]
```diff
@@ -100,25 +101,18 @@ export default class ThumbnailSlider extends EventSubscriber {
...
+     * size of thumbnails in slider (height + margin)
      * @memberof ThumbnailSlider
      * @type {number}
      */
-    thumbnails_start_index = 0;
+    thumbnail_size = 90;
 
     /**
-     * the end index of the displayed thumbnails
+     * height of slider, so we which thumbs have scrolled into view
```
Comment (jburel) [inline src/app/thumbnail-slider.js]: missing verb

File: src/app/thumbnail-slider.js [hunk 3]
```diff
@@ -141,22 +135,25 @@ export default class ThumbnailSlider extends EventSubscriber {
...
+                [IMAGE_VIEWER_RESIZE,
+                    (params={}) => this.resizeViewer(params)]];
 
     /**
      * @constructor
      * @param {Context} context the application context (injected)
      * @param {BindingEngine} bindingEngine the BindingEngine (injected)
      */
-    constructor(context, element, bindingEngine) {
+    constructor(context, element, bindingEngine, taskQueue) {
```
Comment (jburel) [inline src/app/thumbnail-slider.js]: Declare new parameter

File: src/app/thumbnail-slider.js [hunk 4]
```diff
@@ -314,94 +329,105 @@ export default class ThumbnailSlider extends EventSubscriber {
...
+            '@id': id,
+            Name: 'image: ' + id
+        }));
+        this.addThumbnails(to_add, 0);
+    }
+
+    /**
+     * Loads unloaded thumbnails that scrolled into the thumbnail panel
+     *
+     * @param {number} scrollTop Current scroll position of the panel
```
Comment (jburel) [inline src/app/thumbnail-slider.js]: declare parameter

File: src/app/thumbnail-slider.js [hunk 5]
```diff
@@ -480,35 +510,59 @@ export default class ThumbnailSlider extends EventSubscriber {
     }
 
     /**
-     * Adds thumbnails to then be loaded.
+     * Adds thumbnails with Image IDs to this.thumbnails array.
+     * We create a new list from old, replacing the thumbnails (instead of
+     * modifying the old list) to make sure the changes is observed and
```
Comment (jburel) [inline src/app/thumbnail-slider.js]: is/are

File: src/app/thumbnail-slider.js [hunk 6]
```diff
@@ -480,35 +510,59 @@ export default class ThumbnailSlider extends EventSubscriber {
...
-                if (!skip_decrement) this.thumbnails_start_index--;
+                return thumb;
+            }
+        });
+    }
+
+    /**
+     * Adds thumbnails with Image IDs to this.thumbnails array.
+     * We create a new list from old, replacing the thumbnails (instead of
+     * modifying the old list) to make sure the changes is observed and
```
Comment (jburel) [inline src/app/thumbnail-slider.js]: is/are

File: src/app/thumbnail-slider.js [hunk 7]
```diff
@@ -480,35 +510,59 @@ export default class ThumbnailSlider extends EventSubscriber {
...
+     * UI updates.
+     *
+     * @param {Array.<Object>} thumbnails the thumbnails to be added
```
Comment (jburel) [inline src/app/thumbnail-slider.js]: thumbnails not a parameter

**Comment by jburel** [comment]
Testing in IE 11 
it works nicely too

**Pull request merged**